src/Language/PureScript/Sugar/BindingGroups.hs

```diff
@@ -67,12 +67,12 @@ createBindingGroups moduleName = mapM f <=< handleDecls
   handleDecls :: [Declaration] -> m [Declaration]
   handleDecls ds = do
     let values = mapMaybe (fmap (fmap extractGuardedExpr) . getValueDeclaration) ds
-        kindDecls = fmap (,True) $ filter isKindDecl ds
-        dataDecls = fmap (,False) $ filter (\a -> isDataDecl a || isExternDataDecl a || isTypeSynonymDecl a || isTypeClassDecl a) ds
+        kindDecls = fmap (,True) $ filter (\a -> isKindDecl a || isExternDataDecl a) ds
```

Comment:
`foreign import data` declarations are a bit awkward because they are both a data declaration and a kind declaration at the same time; is it safe to remove them from `dataDecls`? I wonder if in some cases they need to be included in the `dataBindingGroupDecls`. For example, on e56ebda3
```
module Main where

foreign import data A :: Type
data B (x :: A)
```
currently results in
```
at src/Main.purs:4:14 - 4:15 (line 4, column 14 - line 4, column 15)

  Unknown type Main.A

while inferring the kind of A
while checking that type A
  has kind Type
in type constructor B
```

Reply:
Good point! I'll add this test.

Reply:
This happens regardless, so I think there's something else going on here.

Reply:
This is fixed, and it was an issue with the boolean signature check. Boolean logic is not not hard.

Reply:
Can you explain why this was manifesting as an unknown name error? Was the declaration being dropped by this designating step or something?

Reply:
Sure. Dependencies have a name and a boolean flag, which is whether or not it's the kind signature or not. As the comment states, this is somewhat tricky. When something has a kind signature, in _general_ we should depend on only that for kind checking. However, if we are _defining_ a kind signature, and the dependency is a type synonym, we need to depend on the actual type synonym declaration so it can be expanded (gross, I know). Foreign data kinds are really kind declarations. Or perhaps, a foreign data declaration should _really_ be:

```
foreign data Foo :: Type -> Type
foreign data Foo
```

It was failing because the predicate in `mkVert` was asking for a dependency on `A` with `False` instead of `True`, which didn't exist. That's because `isSig` for `B` was `False`. That means the declaration for `B` was depending on the declaration for `A`, which doesn't really exist since we are counting foreign data as a kind signature (`True`). The toposort was then putting `B` before `A` because of implementation details, and so `A` was not bound when it checked `B`.

If there is another way to factor this predicate PLEASE TELL ME because it hurts my head.

Reply:
Ah right, thanks! I had a decent idea of the logic behind when we want to depend on a data declaration itself versus just depending on its kind signature from the comment already (thanks for that, by the way), but I didn't realise that `mkVert` was creating a dependency on a nonexistent vertex. As for refactoring, perhaps fewer booleans might help? What if a vertex id was `(ProperName 'TypeName, VertexType)` rather than `(ProperName 'TypeName, Bool)`, where
```
data VertexType
  = VertexDefinition
  | VertexKindSignature
```
Alternatively you could maybe refactor via De Morgan's law?
```
not isSig || not (n `elem` typeSyns) = not (isSig && n `elem` typeSyns)
```
not sure if that actually makes things any clearer, though.

Reply:
Alright, I've taken out the bool and made it at least reflect the comment better.

Reply:
Looks great!